<launch>
  <!-- CORRIDOR EXPLORATION -->
  <node pkg="robot_controller"
        type="local_map_wall_avoider.py"
        name="local_map_wall_avoider"
        output="screen">
    <!-- topics -->
    <param name="map_topic" value="/map/local_map/obstacle" />
    <param name="cmd_vel_topic" value="/cmd_vel" />
    <param name="turn_direction" value="LEFT" />
    <!-- behaviour params -->
    <param name="forward_speed" value="0.25" />
    <param name="rotation_speed" value="0.4" />
    <param name="rotation_angle" value="1.5708" />  <!-- ~90 deg -->
    <param name="probe_angle" value="1.5708" />      <!-- 45 deg -->
    <param name="safe_distance" value="0.5" />          <!-- upper bound -->
    <param name="max_scan_distance" value="0.7" />
    <param name="min_turn_distance" value="0.05" />      <!-- lower bound -->
    <param name="lateral_width" value="0.15" />
    <param name="obstacle_threshold" value="50" />
    <param name="forward_axis" value="x" />
     <!-- Optional tuning for corridor classification -->
    <param name="side_blocked_threshold" value="0.3" />
    <param name="side_open_threshold" value="0.5" />
  </node>

     <!-- APRILTAG APPROACHING -->
   <node pkg="robot_controller"
        type="apriltag_approaching.py"
        name="apriltag_approach"
        output="screen">
    <!-- Topics -->
    <param name="tag_topic"        value="/tag_detections"/>
    <param name="cmd_vel_topic"    value="/cmd_vel"/>
    <!-- Target pose -->
    <param name="target_distance"  value="0.32"/> <!-- desired z distance to tag [m] -->
    <param name="target_sideways"  value="0.0"/>
    <param name="target_id"        value="-1"/> <!-- -1: follow any tag, else specific ID -->
    <param name="advance_distance" value= "0"/>
    <param name="advance_speed"    value="0"/>
    <!-- Gains -->
    <param name="k_side"           value="0.8"/> <!-- lateral control (x) -->
    <param name="k_forward"        value="0.8"/> <!-- forward control (z) -->
    <param name="k_yaw"            value="5.0"/> <!-- rotation control (orientation.z) -->
    <!-- Velocity limits -->
    <param name="max_side_vel"     value="0.75"/> <!-- m/s -->
    <param name="max_forward_vel"  value="0.60"/> <!-- m/s -->
    <param name="max_yaw_vel"      value="3.80"/> <!-- rad/s -->
    <!-- Tolerances -->
    <param name="tol_x"            value="0.01"/> <!-- m, sideways alignment -->
    <param name="tol_z"            value="0.02"/> <!-- m, distance alignment -->
    <param name="tol_yaw"          value="0.02"/> <!-- unitless, orientation.z -->
    <!-- Misc -->
    <param name="timeout"          value="0.5"/> <!-- s, lose tag after this -->``
    <param name="control_rate"     value="20.0"/> <!-- Hz -->
  </node>

    <!-- SEMANTIC MAPPER-->
    <node pkg="semantic_mapping"
            type="semantic_mapper_node.py"
            name="semantic_mapper_node"
            output="screen">
        <!-- Topics -->
        <param name="yolo_topic"          value="/yolo_rs/xy"/>
        <param name="pose_topic"          value="/orb_slam3/pose_corrected"/>
        <!-- Sampling Window -->
        <param name="initial_duration"    value="1.0"/> 
        <param name="max_duration"        value="2.0"/> 
    </node>
    
    <!-- HISTORICAL MAPPER-->
    <node pkg="semantic_mapping"
            type="historical_mapping_node.py"
            name="historical_mapping_node"
            output="screen">
    </node>

    <!-- SEMANTIC VISUALIZER for JUNCTION -->
    <node pkg="semantic_mapping"
        type="semantic_map_visualizer.py"
        name="semantic_map_visualizer"
        output="screen">
    <param name="frame_id" value="world" />
    <param name="check_rate" value="1.0" />
    <param name="semantic_map_path"
           value="$(find semantic_mapping)/maps/semantic_map.json" />
   </node>

    <!-- SEMANTIC VISUALIZER for STORE COORDINATES -->
    <node pkg="semantic_mapping"
        type="storing_coordinate.py"
        name="storing_coordinate_node"
        output="screen">
    <param name="yolo_topic" value="/yolo_rs/xy" />
    <param name="max_radius" value="0.6" />
    <param name="trigger_topic" value="/successful_mark_entering_store"/>
   </node>

   <!-- PRE ENTERING -->
   <node pkg="robot_controller"
        type="pre_entering_node.py"
        name="pre_entering_node"
        output="screen">
    <param name="forward_distance" value="0.5" />
    <param name="forward_speed" value="0.25" />
    <param name="rotation_speed" value="0.5"/>
    <param name="rotation_angle_deg" value="90"/>
    </node>

    <!-- ENTERING STORE-->
    <node pkg="robot_controller"
        type="entering_store_enhanced_node.py"
        name="entering_store_node"
        output="screen">
    <param name="yolo_topic" value="/yolo_rs/xy"/>
    <param name="scan_angle_deg" value="45"/>
    <param name="full_angle_deg" value="90"/>
    <param name="image_width" value="960" />
    <param name="image_height" value="540" />
    <param name="rotation_speed" value="0.5"/>
    <param name="forward_speed" value="0.5"/>
    <param name="side_speed" value="0.3"/>
    <param name="center_tolerance" value="20"/>
    </node>
    
    <!-- MAIN CONTROLLER -->
    <node pkg="robot_controller" type="main_controller.py" name="main_controller_node" output="screen">
    <param name="post_grasp_turn_degrees" value="90"/>
    <param name="tag_lost_far_threshold" value="0.4"/>
    </node>
    
    <!-- DECISION MAKER -->
    <node pkg="robot_controller" type="decision_makers.py" name="decision_makers_node" output="screen" />
    
    <!-- APPROACHING AND GRASPING -->
    <node pkg="robot_controller" type="object_approaching_grasping_enhanced.py" name="object_approaching_grasping_node" output="screen">
    <param name="search_backup_dist" value="0.05"/>
    </node>

    <!-- OBJECT PICKUP CONTROLLER -->
    <node pkg="robot_controller" type="object_pickup_controller.py" name="object_pickup_controller" output="screen">
        <param name="gate_topic" value="/activate_object_pickup"/>
        <param name="costmap_topic" value="/map/local_map/obstacle"/>
        <param name="cmd_vel_topic" value="/cmd_vel"/>
        <param name="rotate_degrees" value="220.0"/>
        <param name="forward_speed" value="0.3"/>
        <param name="rotate_speed" value="0.6"/>
        <param name="publish_hz" value="20"/>
    </node>


</launch>
